Add generated macro and DT macro to count the number of actual peripheral on bus

swift-tk opened the pull request:
## Summary
This PR adds generic devicetree support to count bus-connected device descendants under a controller node, without introducing controller-specific special handling in edtlib.

## Problem
For hierarchical devicetree layouts, direct-child counting is not enough. Real peripheral nodes may be nested under structural/container nodes, and some child nodes may themselves be bus controllers.

Using only direct children can undercount real devices and overfit logic to specific controller implementations.
Take the following device tree as an example. It has 6 peripherals, but with normal child counting, there is 7, not to mention some of them aren't actual peripherals but fake/intermediate nodes. 

```
		test_children_on_bus: controller@ccccdddd {
			compatible = "vnd,controller";
			reg = <0xccccdddd 0x1000>;
			#address-cells = <1>;
			#size-cells = <1>;

			/* Child number 1 of the controller */
			test_child_on_bus_a: test-child-a@0 {
				compatible = "vnd,device";
				reg = <0x0 0x100>;
				status = "okay";
			};

			/* Child number 2 of the controller */
			test_child_on_bus_b: test-child-b@100 {
				compatible = "vnd,device";
				reg = <0x100 0x100>;
				status = "okay";
			};

			/* Child number 3 of the controller */
			test_child_on_bus_c: test-child-c@200 {
				compatible = "vnd,device";
				reg = <0x200 0x100>;
				status = "disabled";
			};

			test_child_on_bus_d: test-child-on-bus-d {
				#address-cells = <1>;
				#size-cells = <1>;
				ranges;

				/* Child number 4 of the controller */
				test_child_dd: test-child-dd@300 {
					compatible = "vnd,device";
					reg = <0x300 0x100>;
					status = "okay";
				};

				/* Child number 5 of the controller */
				test_child_ee: test-child-ee@400 {
					compatible = "vnd,device";
					reg = <0x400 0x100>;
					status = "disabled";
				};
			};

			/* Child number 6 of the controller */
			test_child_on_bus_ext: test-child-on-bus-ext@500 {
				compatible = "vnd,device-extender";
				reg = <0x500 0x300>;
				ranges = <0x0 0x600 0x200>;
				#address-cells = <1>;
				#size-cells = <1>;

				/* Child number 1 of the extender */
				test_child_a_on_ext: test-child-a-on-ext@0 {
					compatible = "vnd,device";
					reg = <0x0 0x100>;
					status = "okay";
				};

				/* Child number 2 of the extender */
				test_child_b_on_ext: test-child-b-on-ext@100 {
					compatible = "vnd,device";
					reg = <0x100 0x100>;
					status = "disabled";
				};
			};

			/* Not a phyiscal child node of the controller */
			controller_dev_setting_a: setting-a {
				compatible = "vnd,controller-device-setting";
				val1 = <1>;
			};

			/* Not a phyiscal child node of the controller */
			controller_dev_setting_b: setting-b {
				compatible = "vnd,controller-device-setting";
				val1 = <2>;
			};
		};
``` 

## Solution
Adds resolved Node.on_bus usage for generation logic, validating binding on-bus against the parent controller’s bus types.
Generates descendant-on-bus count macros in [gen_defines.py](https://file+.vscode-resource.vscode-cdn.net/Users/swift/.vscode/extensions/openai.chatgpt-0.4.71-darwin-x64/webview/#).
Exposes new public macros:
DT_DESCENDANT_NUM_ON_BUS(node_id, bus)
DT_DESCENDANT_NUM_ON_BUS_STATUS_OKAY(node_id, bus)
DT_INST_DESCENDANT_NUM_ON_BUS(inst, bus)
DT_INST_DESCENDANT_NUM_ON_BUS_STATUS_OKAY(inst, bus)
Updates docs and tests accordingly.

### Counting Semantics
Counts nodes whose resolved on_bus matches the queried bus.
Walks descendants (not only direct children).
Stops recursion at child bus-controller nodes.
A child bus-controller node can be counted for the parent bus (if it is on that bus), but devices behind that child bus are excluded from the parent’s count.
_STATUS_OKAY variants apply status filtering.

### Test Coverage
The new devicetree API test topology validates:

Nested devices under non-bus container nodes are counted.
Child bus-controller nodes are counted on the parent bus when applicable.
Devices behind child buses are excluded from parent-bus totals.
Status-okay counts are correct.
Example validated totals:

Parent controller on generic: total 6, status-okay 4
Extender node on generic: total 2, status-okay 1

## Practical Use Cases
Useful for SPI/MSPI and similar multi-peripheral controllers where DTS may include:

Structural/grouping nodes (not actual peripherals),
Stacked-die or wrapper nodes,
Vendor-specific timing/config nodes.
This keeps hierarchy flexible while preserving generic, predictable bus-device counting.

Note that the example is to show one of the possible ways that the DTS could be written without putting specific restrictions in python scripts. 
A few things to note:
- The `ambiq,timing` nodes do not represent an actual peripheral device that are on mspi bus as they are just "grouped" controller specific timing properties by the actual peripheral device i.e. psram and flash. Again, it is one of the possible ways to write down vendor specific controller properties that varies based on the connected peripheral device.
- The actual device nodes in this example have a reg property and its value only represents the sequence as DT allows non consecutive increment. The devices are : S70KL1282_1, S70KL1282_2, is25wx064
- S70KL1282 is a stacked die device which has two spi interfaces to the mspi controller. The node is not on the mspi bus but only its two direct children. **Whether or not they are on mspi bus, it shall be represented by the on_bus key in their compatible bindings**. Again, it is one of the possible ways to represent a stacked die device, but I feel it has the correct hierarchy one would expect. Additionally, there could be some device specific properties such as reset-gpio that are not part of the spi device properties. In this case, the parent holds all the common properties of the its children via additional-property: true, but the example could be written in a way such that every children repeats everything that the parent currently has.
```
&mspi0 {
	pinctrl-0 = <&mspi0_default>;
	pinctrl-1 = <&mspi0_sleep>;
	pinctrl-2 = <&mspi0_psram>;
	pinctrl-3 = <&mspi0_flash>;
	pinctrl-names = "default","sleep","psram","flash";
	status = "okay";
	zephyr,pm-device-runtime-auto;

	cmdq-buffer-location = "SRAM_NO_CACHE";
	cmdq-buffer-size = <256>;

	ambiq,hyperram;

	dqs-support;

	psram_timing: ambiq_timing1 {
		compatible = "ambiq,timing";
		ambiq,timing-config-mask = <0x62>;
		ambiq,timing-config = <8 7 1 0 0 3 10>;
	}

	flash_timing: ambiq_timing2 {
		compatible = "ambiq,timing";
		ambiq,timing-config-mask = <0x62>;
		ambiq,timing-config = <0 16 1 0 0 5 20>;
	}

	S70KL1282: S70KL1282 {
		compatible = "mspi-s70kl1282";
		status = "okay";
		mspi-io-mode = "MSPI_IO_MODE_OCTAL";
		mspi-data-rate = "MSPI_DATA_RATE_S_D_D";
		mspi-max-frequency = <125000000>;
		mspi-dqs-enable;
		read-command = <0x20>;
		write-command = <0xA0>;
		command-length = "INSTR_1_BYTE";
		address-length = "ADDR_4_BYTE";
		rx-dummy = <7>;
		tx-dummy = <8>;
		S70KL1282_1: S70KL1282_1@0 {
			compatible = "mspi-s70kl1282-child",
			size = <DT_SIZE_M(64)>;
			reg = <0>;
			status = "okay";
			mspi-hardware-ce-num = <0>;
			xip-config = <1 0 DT_SIZE_M(8) 0>;
		};
		S70KL1282_2: S70KL1282_2@1 {
			compatible = "mspi-s70kl1282-child",
			size = <DT_SIZE_M(64)>;
			reg = <0>;
			status = "okay";
			mspi-hardware-ce-num = <1>;
			xip-config = <1 DT_SIZE_M(8) DT_SIZE_M(8) 0>;
		};
	};
	is25wx064: is25wx064@2 {
		compatible = "mspi-is25xX0xx", "ambiq,mspi-device";
		size = <DT_SIZE_M(64)>;
		reg = <2>;
		status = "okay";
		mspi-max-frequency = <96000000>;
		mspi-io-mode = "MSPI_IO_MODE_OCTAL_1_8_8";
		mspi-data-rate = "MSPI_DATA_RATE_SINGLE";
		mspi-dqs-enable;
		mspi-hardware-ce-num = <2>;
		read-command = <0xCC>;
		write-command = <0x8E>;
		command-length = "INSTR_1_BYTE";
		address-length = "ADDR_4_BYTE"; 
		rx-dummy = <16>;
		tx-dummy = <0>;
		xip-config = <1 0 DT_SIZE_M(8) 0>;
		ce-break-config = <0 0>;
	};
};
```
The controller may use the actual child number to validate against the maximum allowed number of child devices.

swift-tk committed:
scripts: devicetree: add DESCENDANT_NUM_ON_BUS generated macros

Add generated macros for counting nodes on a specific bus under a bus
controller:

- DT_(INST_)DESCENDANT_NUM_ON_BUS()
- DT_(INST_)DESCENDANT_NUM_ON_BUS_STATUS_OKAY()

The count intentionally walks descendants instead of only direct children.
This covers common devicetree structures where helper/container nodes sit
between a controller and the actual on-bus device nodes.

Traversal stops at child bus-controller nodes, so devices behind sub-buses
are not counted in the parent controller's result.

Also add edtlib Node.on_bus resolution, validated against the parent bus
controller bus types, so generation uses the resolved on-bus value.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
tests: dts: cover DESCENDANT_NUM_ON_BUS counting rules

Add devicetree API coverage for DT_(INST_)DESCENDANT_NUM_ON_BUS() and
DT_(INST_)DESCENDANT_NUM_ON_BUS_STATUS_OKAY().

The test topology validates that:

- devices nested below non-bus container nodes are still counted for the
  parent controller bus
- child nodes that are bus controllers are counted for the parent bus
- devices behind child buses are excluded from the parent count
- status-okay filtering matches generated *_STATUS_OKAY macro values

Update overlay comments to document the expected counting behavior.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

decsny reviewed (changes requested):
there is not supposed to be children of bus nodes that are not on the bus. can you point to a specific example of where you think that would be necessary.

swift-tk commented:
> there is not supposed to be children of bus nodes that are not on the bus. can you point to a specific example of where you think that would be necessary.

Even if that's true, it is not enforced in edtlib.py. And I doubt we could impose such restriction...
The following DTS structure would make DT_CHILD_NUM* macro no longer accurate on describing the actual number of peripheral nodes. 

```
			test_children_on_bus: controller@77778888 {
			compatible = "vnd,controller";
			reg = <0x77778888 0x1000>;
			#address-cells = <1>;
			#size-cells = <0>;
			test_child_on_bus_a: test-child-a@0 {
				compatible = "vnd,device";
				reg = <0>;
				status = "okay";
			};
			test_child_on_bus_b: test-child-b@1 {
				compatible = "vnd,device";
				reg = <1>;
				status = "okay";
			};
			test_child_on_bus_c: test-child-c@2 {
				compatible = "vnd,device";
				reg = <2>;
				status = "disabled";
			};
			test_child_on_bus_d: test-child-on-bus-d {
				#address-cells = <1>;
				#size-cells = <0>;
				test_child_dd: test-child-dd@3 {
					compatible = "vnd,device";
					reg = <3>;
					status = "okay";
				};
				test_child_ee: test-child-ee@4 {
					compatible = "vnd,device";
					reg = <4>;
					status = "okay";
				};
			};
		};
```
Additionally, there were proposals on handling vendor specific hardware settings on the controller node by introducing specific nodes that would not be on the bus as the nodes aren't actual peripheral devices.
  
```
			test_children_on_bus: controller@77778888 {
			compatible = "vnd,controller";
			reg = <0x77778888 0x1000>;
			#address-cells = <1>;
			#size-cells = <0>;
			test_child_on_bus_a: test-child-a@0 {
				compatible = "vnd,device";
				reg = <0>;
				status = "okay";
			};
			test_child_on_bus_b: test-child-b@1 {
				compatible = "vnd,device";
				reg = <1>;
				status = "okay";
			};
			controller_setting_a: controller-setting-a {
				compatible = "vnd,controller-setting";
				val1 = <1>;
			};
			controller_setting_b: controller-setting-b {
				compatible = "vnd,controller-setting";
				val1 = <2>;
			};
		};
```
The Node.on_buses requirement seems pretty loose, shouldn't it be looking at Binding.on_bus first?
```
def on_buses(self) -> list[str]:
        "See the class docstring"
        bus_node = self.bus_node
        return bus_node.buses if bus_node else []
```
I see that I have to change edtlib.py to get what I actually want....

swift-tk commented:
How about I just add a different proprety to class Node below to keep that information from binding and then use it in gen_defines.py
```
    @property
    def on_bus(self) -> str:
        "See the class docstring"
        bus_node = self.bus_node
        if self._binding and self._binding.on_bus is not None:
            if bus_node and self._binding.on_bus in bus_node.buses:
                return self._binding.on_bus
            else
                return None
```

decsny commented:
> >;

Yes, I looked through the PR, I can see the test overlay you added. Now what I am asking for, if it was not implied, is a real example, not an abstract one.

decsny commented:
You are still not answering my question. You are just being hand wavey about okay MSPI can have this and that and whatever, and then showing the abstract DTS again which I said I don't want to see. Why don't you show the actual DTS as you envision or have for this MSPI to be clear of what you are talking about.

swift-tk commented:
> You are still not answering my question. You are just being hand wavey about okay MSPI can have this and that and whatever, and then showing the abstract DTS again which I said I don't want to see. Why don't you show the actual DTS as you envision or have for this MSPI to be clear of what you are talking about.

I'm advocating this for a more generic use not just for MSPI as it could fix the potential loop hole for 1 to many controller-device relations.
Anyways, here you go, S70KL1282 is a stacked die device with two mspi interface. There are three actual peripheral device for the controller not 2, because there is three hardware CE. The timing nodes should not be counted as a peripheral devices.
```
&mspi0 {
	pinctrl-0 = <&mspi0_default>;
	pinctrl-1 = <&mspi0_sleep>;
	pinctrl-2 = <&mspi0_psram>;
	pinctrl-3 = <&mspi0_flash>;
	pinctrl-names = "default","sleep","psram","flash";
	status = "okay";
	zephyr,pm-device-runtime-auto;

	cmdq-buffer-location = "SRAM_NO_CACHE";
	cmdq-buffer-size = <256>;

	ambiq,hyperram;

	dqs-support;

	psram_timing: ambiq_timing1 {
		compatible = "ambiq,timing";
		ambiq,timing-config-mask = <0x62>;
		ambiq,timing-config = <8 7 1 0 0 3 10>;
	}

	flash_timing: ambiq_timing2 {
		compatible = "ambiq,timing";
		ambiq,timing-config-mask = <0x62>;
		ambiq,timing-config = <0 16 1 0 0 5 20>;
	}

	S70KL1282: S70KL1282 {
		compatible = "s70kl1282";
		status = "okay";
		mspi-io-mode = "MSPI_IO_MODE_OCTAL";
		mspi-data-rate = "MSPI_DATA_RATE_S_D_D";
		mspi-max-frequency = <125000000>;
		mspi-dqs-enable;
		read-command = <0x20>;
		write-command = <0xA0>;
		command-length = "INSTR_1_BYTE";
		address-length = "ADDR_4_BYTE";
		rx-dummy = <7>;
		tx-dummy = <8>;
		S70KL1282_1: S70KL1282_1@0 {
			compatible = "mspi-s70kl1282",
			size = <DT_SIZE_M(64)>;
			reg = <0>;
			status = "okay";
			mspi-hardware-ce-num = <0>;
			xip-config = <1 0 DT_SIZE_M(8) 0>;
		};
		S70KL1282_2: S70KL1282_2@1 {
			compatible = "mspi-s70kl1282",
			size = <DT_SIZE_M(64)>;
			reg = <0>;
			status = "okay";
			mspi-hardware-ce-num = <1>;
			xip-config = <1 DT_SIZE_M(8) DT_SIZE_M(8) 0>;
		};
	};
	is25wx064: is25wx064@2 {
		compatible = "mspi-is25xX0xx", "ambiq,mspi-device";
		size = <DT_SIZE_M(64)>;
		reg = <2>;
		status = "okay";
		mspi-max-frequency = <96000000>;
		mspi-io-mode = "MSPI_IO_MODE_OCTAL_1_8_8";
		mspi-data-rate = "MSPI_DATA_RATE_SINGLE";
		mspi-dqs-enable;
		mspi-hardware-ce-num = <2>;
		read-command = <0xCC>;
		write-command = <0x8E>;
		command-length = "INSTR_1_BYTE";
		address-length = "ADDR_4_BYTE"; 
		rx-dummy = <16>;
		tx-dummy = <0>;
		xip-config = <1 0 DT_SIZE_M(8) 0>;
		ce-break-config = <0 0>;
		ambiq,timing-config-mask = <0x62>;
		ambiq,timing-config = <0 16 1 0 0 5 20>;
	};
};
```

swift-tk commented on scripts/dts/gen_defines.py:
```diff
@@ -1003,6 +1003,42 @@ def write_global_macros(edt: edtlib.EDT):
             out_define(
                 f"DT_COMPAT_{str2ident(compat)}_BUS_{str2ident(bus)}", 1)
 
+    bus_by_id = {}
+    for n in edt.nodes:
+        if n.bus_node is not None:
+            bus_by_id[n.bus_node.z_path_id] = n.bus_node
+
+    def _iter_descendants(node):
+        for c in node.children.values():
+            yield c
+            yield from _iter_descendants(c)
+
+    # For each bus controller, and for each bus type it exposes, count the
+    # descendant nodes that are on that bus but have no descendant also on
+    # that same bus (i.e. count only the bus-leaf nodes).
+    for bus_id, bus_node in bus_by_id.items():
+
+        for bus in bus_node.buses:
+            bus_ident = str2ident(bus)
+
+            count = 0
+            count_ok = 0
+            for d in _iter_descendants(bus_node):
+                if bus in d.on_buses:
```
I will get on improve the coding and variable naming if the idea is generally ok.

swift-tk commented on scripts/dts/gen_defines.py:
```diff
@@ -1003,6 +1003,42 @@ def write_global_macros(edt: edtlib.EDT):
+            count_ok = 0
+            for d in _iter_descendants(bus_node):
+                if bus in d.on_buses:
+                    # If any descendant of 'd' (within the controller subtree)
+                    # is also on this bus, then 'd' should NOT be counted.
```
For example, in the actual device tree `S70KL1282: S70KL1282` would not counted as one device on mspi bus but two because it has two MSPI interfaces that may or may not have different properties of the same types i.e. "mspi-device", So `S70KL1282: S70KL1282`'s child would need to be counted as their binding would have on-bus: mspi.
"mspi-s70kl1282" binding may look like:
```
compatible: "mspi-s70kl1282"

child-binding:
  properties:
    mspi-io-mode:
      type: string
      required: true

    mspi-data-rate:
      type: string
      required: true
```
But this is part where I need binding's from parent to be merged down to its children even when they have a compatible string. It is what I'm proposing in https://github.com/zephyrproject-rtos/zephyr/pull/96083, but we don't need to discuss that here.

swift-tk commented:
> There are so many problems with this DT that you shared, that I am not even going to get into right now. But focusing on the main topic which is this bus children count - Are you saying the macro should return `3` for this mspi0 because it should be counting children of children (of children) which are also on the "mspi" bus? If so, this is totally wrong, because grandchildren are not on the bus of their grandparent, they are on the bus of their parent. A node can be for example a bus extender which for example could be a slave on one spi bus and a master of another.

No, the idea is to only count the number of leaf children with on-bus: mspi in their binding under the controller node with bus: mspi in its binding. Any intermediate parent nodes whether are have on-bus:mspi or not in their binding should not be counted as their are not the actual peripheral device. 

I get what you are saying with Node.on_buses is used in the gen_define.py, I realize that I need to add another property to keep Binding.on_bus info in the nodes instead and use it in gen_define.py.

> grandchildren are not on the bus of their grandparent, they are on the bus of their parent. 

The problem I see in the logic here is that, the Node.on_buses ignore the Binding.on_bus of its children. If the intermediate parent is not a bus node, how could its children on their parent bus? Where in my case, via Binding.on_bus, it explicitly stated what bus it should be on.

> A node can be for example a bus extender which for example could be a slave on one spi bus and a master of another.

Yes, that's indeed a problem, then we would need to check Node.buses and Binding.on_bus in here, if the extender also have an interface(i.e. bus: spi), it should be counted otherwise it should not.
has_desc_on_bus = any(
                        bus in dd.on_buses for dd in _iter_descendants(d)
                    )

decsny commented:
> > There are so many problems with this DT that you shared, that I am not even going to get into right now. But focusing on the main topic which is this bus children count - Are you saying the macro should return `3` for this mspi0 because it should be counting children of children (of children) which are also on the "mspi" bus? If so, this is totally wrong, because grandchildren are not on the bus of their grandparent, they are on the bus of their parent. A node can be for example a bus extender which for example could be a slave on one spi bus and a master of another.
> 
> No, the idea is to only count the number of leaf children with on-bus: mspi in their binding under the controller node with bus: mspi in its binding. Any intermediate parent nodes whether they have on-bus:mspi or not in their binding should not be counted as their are not the actual peripheral device. But like you said, extender with bus:mspi should be counted and not its grandchildren.

I mean this is just not right, you should be putting everything that is on the bus as a direct children and the CS should be given in the reg property.  In fact, I am questioning why we even added an `mspi` bus because at the end of the day these device are all spi devices and should not be described any differently as HW than as if they were on a regular SPI controller's bus. The same DT node for a SPI device node should be able to be put on a SPI or MSPI controller, IMO. I just was in the arch WG talking about how we need to put timing characteristics and things like this on the SPI device nodes in DT and it seems to have got approved, I'm just waiting for merge now, but I see no reason why mspi can't reuse this spi device binding and help adding to it. https://github.com/zephyrproject-rtos/zephyr/pull/87427

> > grandchildren are not on the bus of their grandparent, they are on the bus of their parent.
> 
> The problem I see in the logic here is that, the Node.on_buses ignore the Binding.on_bus of its children. If the intermediate parent is not a bus node, how could its children on their parent bus? Where in my case, via Binding.on_bus, it explicitly stated what bus it should be on.

This is not about the python script.  If you did not already know this, Zephyr does not define what is correct Devicetree syntax or semantics.  Devicetree is a project-neutral standard https://www.devicetree.org/ that has it's own specification upstream of Zephyr.  The specification heavily implies that only direct-parent child relationships define buses, and in fact, the parent-child relationships in DT are mainly for the purpose of describing buses.

> > A node can be for example a bus extender which for example could be a slave on one spi bus and a master of another.
> 
> Yes, that's indeed a problem, then I could first go through all direct children, if there is an extender also have an interface(i.e. bus: spi), it should be counted and not its grandchildren.

I don't understand why you actually have to have grandchildren nodes on the bus of a grandparent in the first place.

swift-tk commented:
I don't see how the intention of this PR is not align with how DT is supposed to be use. If it is not supposed to be used this way, in my examples, they there should be documents and restrictions imposed in the script instead of telling me to just not use it this way.

swift-tk commented:
> What I am saying is that the devices on the bus are still SPI devices on a SPI bus. So in DT I don't know why we need an mspi bus that is distinct from spi bus. Again, DT is not supposed to be coupled to project specific software implementations, it is supposed to be a universal description that makes sense for any project using DT.

Please go ahead and open a issue describing this, lets not go off topic here.

swift-tk commented:
@decsny Please enlighten me on the following.

why Binding.on_bus is not being respected? Clearly, if you have some node that doesn't represent an actual peripheral device under a controller, it can not be assumed to be on the controller's bus. IMO, it would have been justifiable if there were hard restriction in the python script, instead, it just blatantly declare the nodes under a bus node must all be on the bus. If you are using the binding bus key to define a bus node, then why wouldn't one define a peripheral node with binding on-bus key?

mbolivar reviewed (changes requested):
Haven't had time to look in detail, but this requires macros.bnf update at the least.

swift-tk commented:
> Haven't had time to look in detail, but this requires macros.bnf update at the least.

Not exactly sure what is needed to change, could you provide some details? Where is macros.bnf?

swift-tk committed:
doc: dts: clarify DESCENDANT_NUM_ON_BUS macro semantics

Document generated macros used by DT_(INST_)DESCENDANT_NUM_ON_BUS() and
DT_(INST_)DESCENDANT_NUM_ON_BUS_STATUS_OKAY().

Clarify that these macros count descendants on the requested bus under the
controller node, while excluding descendants behind child bus controllers.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk commented:
rebase as unidenifiable CI fail

nordicjm commented on include/zephyr/devicetree.h:
```diff
@@ -669,6 +668,28 @@
 
+/**
+ * @brief Get the number of child nodes of a given node
+ *        which child nodes' on a give bus
```
```suggestion
 *        with child nodes' on a given bus
```

nordicjm commented on include/zephyr/devicetree.h:
```diff
@@ -4089,6 +4111,34 @@
 
+/**
+ * @brief Get the number of child nodes of a given node
+ *        which child nodes' on a give bus
```
as above

swift-tk replied:
done.

nordicjm commented on include/zephyr/devicetree.h:
```diff
@@ -4089,6 +4111,34 @@
+
+/**
+ * @brief Get the number of child nodes of a given node
+ *        which child nodes' on a give bus and their status are okay
```
as above

swift-tk replied:
done.

nordicjm commented on include/zephyr/devicetree.h:
```diff
@@ -669,6 +668,28 @@
+
+/**
+ * @brief Get the number of child nodes of a given node
+ *        which child nodes' on a give bus and their status are okay
```
as above

nordicjm commented on include/zephyr/devicetree.h:
```diff
@@ -4079,6 +4100,7 @@
 
 /**
  * @brief Get the number of child nodes of a given node
+ *        which child nodes' status are okay
```
```suggestion
 *        whose child nodes' status are okay
```

nordicjm commented on tests/lib/devicetree/api/app.overlay:
```diff
@@ -677,6 +677,66 @@
+			reg = <0x77778888 0x1000>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			test_child_on_bus_a: test-child-a@0 {
```
not compliant with coding standards e.g. newlines needed on lines 690, 685, etc. fix in whole PR

swift-tk replied:
done.

nordicjm reviewed (commented):
Seems like a good addition

mbolivar commented:
Thanks for the update!

> Useful scenario could be spi/mspi controller device dts and software multi-peripheral management.

I've taken another look at this and I'd like to understand this a bit better.

Could you please update the issue description to show an example of how this is meant to work, and why this needs to query all descendants, and not just children?

Please **update the issue description** instead of replying to this comment so we have an easy-to-read record of the rationale that won't get lost in the many comments in this thread.

mbolivar commented on scripts/dts/gen_defines.py:
```diff
@@ -1016,6 +1016,44 @@ def write_global_macros(edt: edtlib.EDT):
+        if node.bus_node is not None:
+            bus_by_id[node.bus_node.z_path_id] = node.bus_node
+
+    def _iter_descendants(node):
```
The devicetree.h docstrings are about "children", but in various places this just mentions "descendants".

The commit comment also seems to be using the concepts of "child" and "descendant" interchangeably.

These are not the same idea and the difference is important. I think it is worth using clear notation that uses "children" or "descendants" consistently and correctly in:

- code
- comments
- commit message

mbolivar reviewed (changes requested):
I didn't get a clear answer to my question about if you have read the specification and believe you understand it. Please answer.

I tried to review this PR again but I can't understand it, sorry. I left a few comments on the first commit which explain where my trouble starts.

Can you please try to make the descriptions of the APIs clearer and provide examples as requested?

I also tried to read the pull request description again but I'm not clear on what's wrong or why this is a solution.

Thanks.

swift-tk commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1044,6 +1044,11 @@ class Node:
+      Unlike on_buses that just assign to children from bus parent, Binding.on_bus
+      is respected and checked against parent node's buses values. The attribute is
+      either a bus value in parent nodes's buses or None.
+
```
So the issue is that on_buses blindly inherits whatever value is on the parent node, instead of first verifying that the node’s binding actually defines an on_bus key. With on_bus present, this means “fake” / intermediate nodes that don’t correspond to real hardware end up with None and get excluded from DT_CHILD_NUM_ON_BUS counting. Meanwhile, real devices should be unaffected because their bindings include on_bus when they are bus-attached devices.

swift-tk commented:
@mbolivar @decsny PTAL thanks

fabiobaltieri commented:
@swift-tk FYI dismissing reviews is the assignee responsability, I'd normally block this, that said I see Declan hasn't been active in the project for months so I think it's fair to dismiss the review, but please in the future reach out to the assignee or the release team (#release on chat.zephyrproject.org) rather than doing it directly.
